Add flop counter utility

Chillee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95901
* #95854
* __->__ #95751
* #95355

Overall, an example usage. Note that this *also* captures backwards FLOPs.
```
import torchvision.models as models
import torch
from torch.utils.flop_counter import FlopCounterMode

inp = torch.randn(1, 3, 224, 224, device='cpu')
mod = models.resnet18()

flop_counter = FlopCounterMode(mod, depth=1)
with flop_counter:
    mod(inp).sum().backward()
```
<img width="326" alt="image" src="https://user-images.githubusercontent.com/6355099/222023068-3491e405-f195-4e11-b679-36b19a1380c7.png">

You can control the depth of the module hierarchy with the `depth` attribute (which defaults to 2). For example, if I don't limit it, this is what it outputs.

<img width="366" alt="image" src="https://user-images.githubusercontent.com/6355099/222023306-3d880bb6-f534-4f98-bf10-83c4353acefc.png">

## Other APIs

FlopCounterMode(custom_mapping=...): Allows for custom flop counting functions
FlopCounterMode.get_table(depth=...): Explicitly get the table as a string
FlopCounterMode.flop_counts: Contains the flop information as a Dict[hierarchy: str, Dict[Op, int]]
FlopCounterMode.register_hierarchy(f, name): Allows you to register additional "hierarchies" for a function.

Chillee committed:
Add flop counter utility

[ghstack-poisoned]

Chillee committed:
Update on "Add flop counter utility"




[ghstack-poisoned]

Chillee committed:
Update on "Add flop counter utility"




[ghstack-poisoned]

Chillee committed:
Update on "Add flop counter utility"


Overall, an example usage. Note that this *also* captures backwards FLOPs.
```
import torchvision.models as models
import torch
from torch.utils.flop_counter import FlopCounterMode

inp = torch.randn(1, 3, 224, 224, device='cpu')
mod = models.resnet18()

flop_counter = FlopCounterMode(mod, depth=1)
with flop_counter:
    mod(inp).sum().backward()
```
<img width="326" alt="image" src="https://user-images.githubusercontent.com/6355099/222023068-3491e405-f195-4e11-b679-36b19a1380c7.png">

You can control the depth of the module hierarchy with the `depth` attribute (which defaults to 2). For example, if I don't limit it, this is what it outputs.

<img width="366" alt="image" src="https://user-images.githubusercontent.com/6355099/222023306-3d880bb6-f534-4f98-bf10-83c4353acefc.png">

## Other APIs

FlopCounterMode(custom_mapping=...): Allows for custom flop counting functions
FlopCounterMode.get_table(depth=...): Explicitly get the table as a string
FlopCounterMode.flop_counts: Contains the flop information as a Dict[hierarchy: str, Dict[Op, int]]


[ghstack-poisoned]

Chillee committed:
Update on "Add flop counter utility"


Overall, an example usage. Note that this *also* captures backwards FLOPs.
```
import torchvision.models as models
import torch
from torch.utils.flop_counter import FlopCounterMode

inp = torch.randn(1, 3, 224, 224, device='cpu')
mod = models.resnet18()

flop_counter = FlopCounterMode(mod, depth=1)
with flop_counter:
    mod(inp).sum().backward()
```
<img width="326" alt="image" src="https://user-images.githubusercontent.com/6355099/222023068-3491e405-f195-4e11-b679-36b19a1380c7.png">

You can control the depth of the module hierarchy with the `depth` attribute (which defaults to 2). For example, if I don't limit it, this is what it outputs.

<img width="366" alt="image" src="https://user-images.githubusercontent.com/6355099/222023306-3d880bb6-f534-4f98-bf10-83c4353acefc.png">

## Other APIs

FlopCounterMode(custom_mapping=...): Allows for custom flop counting functions
FlopCounterMode.get_table(depth=...): Explicitly get the table as a string
FlopCounterMode.flop_counts: Contains the flop information as a Dict[hierarchy: str, Dict[Op, int]]


[ghstack-poisoned]

Chillee committed:
Update on "Add flop counter utility"


Overall, an example usage. Note that this *also* captures backwards FLOPs.
```
import torchvision.models as models
import torch
from torch.utils.flop_counter import FlopCounterMode

inp = torch.randn(1, 3, 224, 224, device='cpu')
mod = models.resnet18()

flop_counter = FlopCounterMode(mod, depth=1)
with flop_counter:
    mod(inp).sum().backward()
```
<img width="326" alt="image" src="https://user-images.githubusercontent.com/6355099/222023068-3491e405-f195-4e11-b679-36b19a1380c7.png">

You can control the depth of the module hierarchy with the `depth` attribute (which defaults to 2). For example, if I don't limit it, this is what it outputs.

<img width="366" alt="image" src="https://user-images.githubusercontent.com/6355099/222023306-3d880bb6-f534-4f98-bf10-83c4353acefc.png">

## Other APIs

FlopCounterMode(custom_mapping=...): Allows for custom flop counting functions
FlopCounterMode.get_table(depth=...): Explicitly get the table as a string
FlopCounterMode.flop_counts: Contains the flop information as a Dict[hierarchy: str, Dict[Op, int]]
FlopCounterMode.register_hierarchy(f, name): Allows you to register additional "hierarchies" for a function.


[ghstack-poisoned]

ezyang commented:
Who do you want to do a close code review on this?

yinghai commented on torch/utils/flop_counter.py:
```diff
@@ -0,0 +1,306 @@
+    return flop_count
+
+
+flop_mapping = {
```
Can we also add support to scaled_dot_product_attention since it's more and more used nowadays?

Chillee replied:
Good idea.

ppwwyyxx replied:
And einsum

Chillee replied:
Einsum isn’t needed since it’s a “compositeimplicit” op and decomposes into other operators.

Chillee replied:
I'm gonna do the attention ones in a follow-up PR.

ppwwyyxx replied:
This actually leads to an important design issue: I think there should be an API to return "unsupported" ops. Similar to https://github.com/facebookresearch/fvcore/blob/51092b5515cbb493f73de079743dd6b11cc4bbf1/fvcore/nn/jit_analysis.py#L98

The reason is that users are mainly interacting with high-level modules and ops, and aren't aware of what low-level ops are called. Imagine one day torch adds a new op for transformer, or adds a different implementation of einsum that doesn't decompose: the change should be transparent, but it makes the flop counter become a silent error!

So in fvcore we provide unsupported ops and print a warning about them by default. To make the results more meaningful, we have a list of trivial ops that are always ignored and will not appear in "unsupported ops" (https://github.com/facebookresearch/fvcore/blob/51092b5515cbb493f73de079743dd6b11cc4bbf1/fvcore/nn/jit_analysis.py#L28).

Chillee replied:
@ppwwyyxx Hmm... I think I might just add all the ops as "unsupported ops". 

I think one benefit of putting it in core is that in principle, we should hopefully be able to keep the operator list up to date better.

ngimel replied:
Yeah, impl changes should be caught by tests, we have an einsum test in this PR, but not matmul?

Chillee commented:
@ezyang Not sure I really need a close code review from anybody 🤔 Maybe @ngimel

ppwwyyxx commented on torch/utils/flop_counter.py:
```diff
@@ -0,0 +1,306 @@
+        func_packet = func._overloadpacket
+        if func_packet in self.flop_mapping:
+            flop_count_func = self.flop_mapping[func_packet]
+            args_shape, out_shape = tree_map(get_shape, (args, normalize_tuple(out)))
```
Is it a good idea to always apply get_shape on tensors? For example I wonder if there would be an op that takes a scalar boolean tensor and condition two different behaviors (with different flops) on it.

Chillee replied:
I don't think it's a big deal - if that ever shows up we can just pass through scalar tensors.

lezcano commented on test/test_flop_counter.py:
```diff
@@ -0,0 +1,123 @@
+        mode = FlopCounterMode()
+        with mode:
+            torch.mm(T(4, 5), T(5, 6))
+        # 4 * 6 * 2 * 5 = 216
+        self.assertExpectedInline(get_total_flops(mode), """240""")
```
Odd. 4 * 6 * 2 * 5 = 240. But the number of FLOPS for an `mm` `[m, k] x [k, n]` is `m * n * (2 * k - 1)`, which in this case gives 216.

lezcano commented on torch/utils/flop_counter.py:
```diff
@@ -0,0 +1,306 @@
+    m, k = a_shape
+    k2, n = b_shape
+    assert k == k2
+    # NB(chilli): Should be 2 * k - 1 technically for FLOPs.
+    return m * n * 2 * k
```
Why not doing it right, now that we're at it?

Chillee committed:
Update on "Add flop counter utility"


Overall, an example usage. Note that this *also* captures backwards FLOPs.
```
import torchvision.models as models
import torch
from torch.utils.flop_counter import FlopCounterMode

inp = torch.randn(1, 3, 224, 224, device='cpu')
mod = models.resnet18()

flop_counter = FlopCounterMode(mod, depth=1)
with flop_counter:
    mod(inp).sum().backward()
```
<img width="326" alt="image" src="https://user-images.githubusercontent.com/6355099/222023068-3491e405-f195-4e11-b679-36b19a1380c7.png">

You can control the depth of the module hierarchy with the `depth` attribute (which defaults to 2). For example, if I don't limit it, this is what it outputs.

<img width="366" alt="image" src="https://user-images.githubusercontent.com/6355099/222023306-3d880bb6-f534-4f98-bf10-83c4353acefc.png">

## Other APIs

FlopCounterMode(custom_mapping=...): Allows for custom flop counting functions
FlopCounterMode.get_table(depth=...): Explicitly get the table as a string
FlopCounterMode.flop_counts: Contains the flop information as a Dict[hierarchy: str, Dict[Op, int]]
FlopCounterMode.register_hierarchy(f, name): Allows you to register additional "hierarchies" for a function.


[ghstack-poisoned]

albanD commented on torch/testing/_internal/common_cuda.py:
```diff
@@ -27,8 +27,9 @@
 if TEST_NUMBA:
-    import numba.cuda
-    TEST_NUMBA_CUDA = numba.cuda.is_available()
+    # import numba.cuda
```
This needs to be reverted before merging.

Chillee replied:
oops

albanD commented on torch/utils/flop_counter.py:
```diff
@@ -0,0 +1,306 @@
+    return x
+
+
+# Thanks BingChat!
```
?

albanD commented on torch/utils/flop_counter.py:
```diff
@@ -0,0 +1,306 @@
+        return wrapped_f
+
+    def get_table(self, depth=None):
+        if depth is None:
+            depth = self.depth
+        if depth is None:
+            depth = 999999
```
```suggestion
        depth = depth or self.depth or 999999
```

Chillee replied:
this doesn't work if depth is 0.

albanD commented on torch/utils/flop_counter.py:
```diff
@@ -0,0 +1,306 @@
+
+        return PopState.apply
+
+    def register_hierarchy(self, f, name):
```
What is this doing?

Chillee replied:
Hmm... it's just providing a way for users to register a custom level in the hierarchy. Perhaps it's unneeded tbh - might remove it.

albanD commented on torch/utils/flop_counter.py:
```diff
@@ -0,0 +1,306 @@
+        self.display = display
+        if custom_mapping is None:
+            custom_mapping = {}
+        if mod is not None:
```
You don't actually need the user to pass in the module or use custom Function.
You can use global hooks like `torch.nn.modules.module.register_module_forward_pre_hook()` that will be called for ALL modules.

More generally, you can use `forward_pre`/`forward` hook to do the forward push-pop and the `full_backward_pre_hook`/`full_backward_hook` to do the backward push pop.

ppwwyyxx replied:
I think that will be a good API improvement. But in that case, what shall we do if users call two different models (which might have shared or distinct layers with the same name) in the context? This might get tricky to handle.

Chillee replied:
oh, full_backward_pre_hook is a new API :P

Chillee replied:
In general, I don't think global module hooks make that much sense given that we're trying to register a hierarchy. Concretely, the hook doesn't tell us where in the module hierarchy it lives.

Chillee replied:
Hmm... I needed to do the "clone" thing in order to work around this issu
```
RuntimeError: Output 0 of BackwardHookFunctionBackward is a view and is being modified inplace. This view was created inside a custom Function (or because an input was returned as-is) and the autograd logic to handle view+inplace would override the custom backward associated with the custom Function, leading to incorrect gradients. This behavior is forbidden. You can fix this by cloning the output of the custom Function.
```

How would I do this with `full_backward_pre_hook`?

albanD replied:
> Concretely, the hook doesn't tell us where in the module hierarchy it lives.

But you can use a global stack to track where you are in the hierarchy right now. Like you know what the current Module is and when you enter a new one, make the current its parent and set the new one as the new current.

> Hmm... I needed to do the "clone" thing in order to work around this issu

I'm not sure what you mean by the clone thing. But yes if you have custom Function that is an issue.
If you have time, I think we can rework this to use multi-Tensor hooks instead of custom Function and that will remove this constraint :)

Chillee replied:
If I have a module like `ReLU` where the forwards pass is
```
def forward(self, x):
    return x.relu_()
```

And I install a backwards hook, it throws that error.

albanD commented on torch/utils/flop_counter.py:
```diff
@@ -0,0 +1,306 @@
+def transpose_shape(shape):
+    return [shape[1], shape[0]] + list(shape[2:])
+
+def conv_backward_flop(input_shapes: List[Any], output_shapes: List[Any]) -> int:
```
Should you just pass in the args/kwargs to these functions? That will simplify the logic in here quite a bit no?

Chillee replied:
Hmm... maybe.

albanD commented on torch/utils/flop_counter.py:
```diff
@@ -0,0 +1,306 @@
+        return i.shape
+    return i
+
+def prod(x):
```
from math import prod

albanD reviewed (commented):
There are a lot of public not documented functions ;)

ppwwyyxx commented on test/test_flop_counter.py:
```diff
@@ -0,0 +1,124 @@
+    return torch.utils.flop_counter.FlopCounterMode(*args, **kwargs, display=False)
+
+def get_total_flops(mode):
+    return str(sum([v for _, v in mode.flop_counts["Global"].items()]))
```
It seems the test would be simpler if this function just returns int.

Chillee replied:
assertExpectedInline expects strings.

ppwwyyxx commented on test/test_flop_counter.py:
```diff
@@ -0,0 +1,124 @@
+
+        self.assertExpectedInline(get_total_flops(mode), """5""")
+
+
```
Other cases to test:
* Module A call a submodule `self.subm.method_name()`, instead of `self.subm.__call__`.  It's ambiguous in this case which module (self or self.subm) the flops should attribute to, but please test that the results are at least reasonable.
* Module with scripted submodule: https://github.com/facebookresearch/fvcore/blob/51092b5515cbb493f73de079743dd6b11cc4bbf1/tests/test_flop_count.py#L832-L845
* Module with legacy autograd function: https://github.com/facebookresearch/fvcore/blob/51092b5515cbb493f73de079743dd6b11cc4bbf1/tests/test_flop_count.py#L820-L830  Not sure if this could be supported at all.

Chillee replied:
> Module with scripted submodule:

Unfortunately we can't support hierarchies on scripted modules, since they don't support hooks. I'll add a test for just capturing flops under scripted functions.

> Module with legacy autograd function:

This works totally fine.

ppwwyyxx replied:
Actually, today I just found that autograd function isn't working in the way I expected.

Usually users will put custom cuda kernels, which pytorch knows nothing about, in such autograd function. For example like this https://github.com/facebookresearch/detectron2/blob/e9f7e2ba15abd7badcb05ef6f5076f06b36a9c5b/detectron2/layers/deform_conv.py#L59

It seems in this PR there is no way to register a custom flop counter for custom kernels that are not registered as a pytorch op (e.g. https://github.com/facebookresearch/detectron2/blob/e9f7e2ba15abd7badcb05ef6f5076f06b36a9c5b/detectron2/layers/csrc/vision.cpp#L83-L94).

However, the fvcore flop counter can interpret the autograd function as an atomic unit (instead of entering it, like this PR), so it actually allows users to count custom kernel properly, for example in this test: https://github.com/facebookresearch/fvcore/blob/e8d19df2ffdaf0a78c8d88d6a2522c36b0cacb07/tests/test_flop_count.py#L820-L830

ppwwyyxx commented on torch/utils/flop_counter.py:
```diff
@@ -0,0 +1,306 @@
+
+    def __enter__(self):
+        self.flop_counts.clear()
+        super().__enter__()
```
maybe `return self`? so users can do:
```
with FlopCountingMode(...) as flop:
   do stuff
```

ppwwyyxx commented on torch/utils/flop_counter.py:
```diff
@@ -0,0 +1,306 @@
+
+    """
+    def __init__(self, mod: torch.nn.Module = None, depth: int = 2, display: bool = True, custom_mapping: Dict[Any, Any] = None):
+        self.flop_counts: Dict[Any, Any] = defaultdict(lambda: defaultdict(int))
```
Users want to access this, either for a custom display format, or for aggregating it across a dataset (e.g. https://github.com/facebookresearch/detectron2/blob/38af375052d3ae7331141bc1a22cfa2713b02987/tools/analyze_model.py#L66-L68)

So it's probably better to make it a proper API with documentation.

ngimel commented on test/test_flop_counter.py:
```diff
@@ -0,0 +1,124 @@
+        # c_in = 3
+        # out_image_size * kernel_size * c_out * 2 * c_in
+
+        # NB: I don't think this properly accounts for padding?
```
it does, via output_size. To compute each output value, you need `kernel_size * c_in * c_out` flops. Sometimes in padded cases you could do super inefficient implementation that computes and throws away padded values, but roofline flop counter shouldn't account for implementation inefficiencies

Chillee replied:
Yeah, but doesn't the padding increase output_size even though we don't need to actually compute them?

ngimel replied:
I guess you could claim that, but in reality you are still computing them in implicit-gemm formulation, those 0's are still part of the gemm tile.

ngimel commented on torch/utils/flop_counter.py:
```diff
@@ -0,0 +1,306 @@
+    mm_flops = mm_flop([a_shape, b_shape])
+    batch_size, _ = a_shape
+    _, output_dim = b_shape
+    return mm_flops + batch_size * output_dim
```
wait, why? `mm` flops already account for this?

Chillee replied:
I was just adding the addition flops, although maybe this is unneeded 🤔

Chillee replied:
Decided to remove them for consistency.

ngimel commented on torch/utils/flop_counter.py:
```diff
@@ -0,0 +1,306 @@
+) -> int:
+    """
+    Count flops for convolution. Note only multiplication is
+    counted. Computation for addition and bias is ignored.
```
what addition?

ngimel commented on test/test_flop_counter.py:
```diff
@@ -0,0 +1,124 @@
+        self.assertExpectedInline(get_total_flops(mode), """720""")
+
+        with mode:
+            torch.addmm(T(4, 6), T(4, 5), T(5, 6))
```
add tests with `alpha` and `beta` args

ngimel commented on torch/utils/flop_counter.py:
```diff
@@ -0,0 +1,306 @@
+        func_packet = func._overloadpacket
+        if func_packet in self.flop_mapping:
+            flop_count_func = self.flop_mapping[func_packet]
+            args_shape, out_shape = tree_map(get_shape, (args, normalize_tuple(out)))
```
Non-tensor args for addmm/baddbmm could throw off flop_count functions I think

Chillee replied:
Since if they're 0 we don't actually do the computation?

ngimel replied:
But `get_shape` would still return them unchanged, and then `addmm_flops` will have 5 args instead of 3?

Chillee replied:
I think those are always passed in as kw-only args.

ngimel replied:
Ah cool, yeah docs say kw-only

Chillee committed:
Update on "Add flop counter utility"


Overall, an example usage. Note that this *also* captures backwards FLOPs.
```
import torchvision.models as models
import torch
from torch.utils.flop_counter import FlopCounterMode

inp = torch.randn(1, 3, 224, 224, device='cpu')
mod = models.resnet18()

flop_counter = FlopCounterMode(mod, depth=1)
with flop_counter:
    mod(inp).sum().backward()
```
<img width="326" alt="image" src="https://user-images.githubusercontent.com/6355099/222023068-3491e405-f195-4e11-b679-36b19a1380c7.png">

You can control the depth of the module hierarchy with the `depth` attribute (which defaults to 2). For example, if I don't limit it, this is what it outputs.

<img width="366" alt="image" src="https://user-images.githubusercontent.com/6355099/222023306-3d880bb6-f534-4f98-bf10-83c4353acefc.png">

## Other APIs

FlopCounterMode(custom_mapping=...): Allows for custom flop counting functions
FlopCounterMode.get_table(depth=...): Explicitly get the table as a string
FlopCounterMode.flop_counts: Contains the flop information as a Dict[hierarchy: str, Dict[Op, int]]
FlopCounterMode.register_hierarchy(f, name): Allows you to register additional "hierarchies" for a function.


[ghstack-poisoned]

albanD commented on torch/utils/flop_counter.py:
```diff
@@ -0,0 +1,323 @@
+    aten.convolution_backward: conv_backward_flop,
+}
+
+def normalize_tuple(x):
```
Please document public API or make them private if they shouldn't be public.

Chillee replied:
ah that's what you meant

Chillee replied:
I can also just set `__all__` right?

albanD replied:
Yes but then you'll see that you can't have these functions that look public and are not in `__all__`. So you will have to prepend them with `_` anyways.

Chillee replied:
Why will I see? It seems to work for me.

albanD reviewed (changes requested):
I need to upgrade the public API script again to pick these up...

Chillee committed:
Update on "Add flop counter utility"


Overall, an example usage. Note that this *also* captures backwards FLOPs.
```
import torchvision.models as models
import torch
from torch.utils.flop_counter import FlopCounterMode

inp = torch.randn(1, 3, 224, 224, device='cpu')
mod = models.resnet18()

flop_counter = FlopCounterMode(mod, depth=1)
with flop_counter:
    mod(inp).sum().backward()
```
<img width="326" alt="image" src="https://user-images.githubusercontent.com/6355099/222023068-3491e405-f195-4e11-b679-36b19a1380c7.png">

You can control the depth of the module hierarchy with the `depth` attribute (which defaults to 2). For example, if I don't limit it, this is what it outputs.

<img width="366" alt="image" src="https://user-images.githubusercontent.com/6355099/222023306-3d880bb6-f534-4f98-bf10-83c4353acefc.png">

## Other APIs

FlopCounterMode(custom_mapping=...): Allows for custom flop counting functions
FlopCounterMode.get_table(depth=...): Explicitly get the table as a string
FlopCounterMode.flop_counts: Contains the flop information as a Dict[hierarchy: str, Dict[Op, int]]
FlopCounterMode.register_hierarchy(f, name): Allows you to register additional "hierarchies" for a function.


[ghstack-poisoned]

albanD reviewed (approved):
Ok if CI is ok!

Chillee commented:
@pytorchbot merge

danthe3rd commented on torch/utils/flop_counter.py:
```diff
@@ -0,0 +1,325 @@
+            def forward(ctx, *args):
+                assert(self.parents[-1] == name)
+                self.parents.pop()
+                args = tree_map(lambda x: x.clone() if isinstance(x, torch.Tensor) else x, args)
```
quick question: why are we calling `clone()` here?

Chillee replied:
If the output of an autograd.Function is a view of the input, there are some restrictions on whether you can use inplace operations or not.
